CRVResponse/src/CrvWaveformsGenerator_module.cc

```diff
@@ -187,76 +186,104 @@ namespace mu2e
-          std::vector<art::Ptr<CrvStep> > stepVector;
-          std::set<art::Ptr<CrvStep> >::iterator stepIter;
-          for(stepIter=steps.begin(); stepIter!=steps.end(); stepIter++) stepVector.push_back(*stepIter);
-
-          //find the most likely SimParticle
-          //if no SimParticle was recorded for this single waveform, then it was caused either by noise hits (if the threshold is low enough),
-          //or is the tail end of the peak. in that case, _simparticle will be null (set by the default constructor of art::Ptr)
-          art::Ptr<SimParticle> simParticle;
-          std::map<art::Ptr<SimParticle>,int >::iterator simparticleIter;
-          int simparticleCount=0;
-          for(simparticleIter=simparticles.begin(); simparticleIter!=simparticles.end(); simparticleIter++)
-          {
-            if(simparticleIter->second>simparticleCount)
+            //loop through the steps to fill the single waveform
+            std::vector<art::Ptr<CrvStep> > stepVector;
+            std::set<art::Ptr<CrvStep> >::iterator stepIter;
+            for(stepIter=steps.begin(); stepIter!=steps.end(); stepIter++) stepVector.push_back(*stepIter);
+
+            //find the most likely SimParticle
+            //if no SimParticle was recorded for this single waveform, then it was caused either by noise hits (if the threshold is low enough),
+            //or is the tail end of the peak. in that case, _simparticle will be null (set by the default constructor of art::Ptr)
+            art::Ptr<SimParticle> simParticle;
+            std::map<art::Ptr<SimParticle>,int >::iterator simparticleIter;
+            int simparticleCount=0;
+            for(simparticleIter=simparticles.begin(); simparticleIter!=simparticles.end(); ++simparticleIter)
             {
-              simparticleCount=simparticleIter->second;
-              simParticle=simparticleIter->first;
+              if(simparticleIter->second>simparticleCount)
+              {
+                simparticleCount=simparticleIter->second;
+                simParticle=simparticleIter->first;
+              }
             }
-          }
 
-          i--;
-          crvDigiMCCollection->emplace_back(voltages, stepVector, simParticle, digiStartTime, TDC0timeAdjusted, barIndex, SiPM);
+            --i;
+            crvDigiMCCollection->emplace_back(voltages, stepVector, simParticle, digiStartTime, TDC0timeAdjusted, barIndex, SiPM);
+          }
         }
       }
     }
 
     event.put(std::move(crvDigiMCCollection));
   } // end produce
 
+  void CrvWaveformsGenerator::FindChargeClusters(const std::vector<CrvSiPMCharges::SingleCharge> &timesAndCharges,
+                                                   std::vector<ChargeCluster> &chargeClusters,
+                                                   double timeShiftFEB)
+  {
+    for(size_t i=0; i<timesAndCharges.size(); ++i)
+    {
+      //No check whether times are within digitizationStart-_digitizationMargin and digitizationEnd
```

Comment:
it will be more efficient to allocate the size of chargeClusters: chargeClusters.reserve(timesAndCharges.size());

Reply:
I can add a reserve().